src/SessionManager.ts

```diff
@@ -0,0 +1,166 @@
+
+const timeoutBlob = new Blob([`self.onmessage = function(e) {
+  setTimeout(() => self.postMessage(""), e.data);
+};`], { type: 'text/javascript' })
+
+export default class SessionManager {
+  private readonly innerTransportFactory: grpc.TransportFactory
+
+  private client: RobotServiceClient
+
+  private currentSessionID = ''
+  private sessionsSupported: boolean | undefined
+  private heartbeatIntervalMs: number | undefined
+
+  private starting: Promise<void> | undefined
+
+  private startResolve: (() => void) | undefined
+
+  private startReject: ((reason: ServiceError) => void) | undefined
+
+  constructor (serviceHost: string, transportFactory: grpc.TransportFactory) {
+    this.innerTransportFactory = transportFactory
+    this.client = new RobotServiceClient(serviceHost, { transport: transportFactory })
+  }
+
+  get transportFactory () {
+    return (opts: grpc.TransportOptions): grpc.Transport => {
+      return new SessionTransport(opts, this.innerTransportFactory, this)
+    }
+  }
+
+  get sessionID () {
+    return this.currentSessionID
+  }
+
+  private getSessionMetadataInner (): grpc.Metadata {
+    const md = new grpc.Metadata()
+    if (this.sessionsSupported && this.currentSessionID !== '') {
+      md.set('viam-sid', this.currentSessionID)
+    }
+    return md
+  }
+
+  public reset () {
+    if (this.starting) {
+      return
+    }
+    this.sessionsSupported = undefined
+  }
+
+  // Note: maybe support non-worker for foreground presence.
+  private readonly backgroundHeartbeat = true
+
+  private async heartbeat () {
+    if (!this.sessionsSupported || this.currentSessionID === '') {
+      return
+    }
+    while (this.starting) {
+      // eslint-disable-next-line no-await-in-loop
+      await this.starting
+    }
+
+    let worker: Worker | undefined
+    const doHeartbeat = () => {
+      const sendHeartbeatReq = new robotApi.SendSessionHeartbeatRequest()
+      sendHeartbeatReq.setId(this.currentSessionID)
+      this.client.sendSessionHeartbeat(sendHeartbeatReq, new grpc.Metadata(), (err) => {
+        if (err) {
+          if (ConnectionClosedError.isError(err)) {
+            /*
+             * We assume the connection closing will cause getSessionMetadata to be
+             * called again by way of a reset.
+             */
+            this.reset()
+            return
+          }
+          // Otherwise we want to continue in case it was just a blip
+        }
+        if (worker) {
+          worker.postMessage(this.heartbeatIntervalMs)
+        } else {
+          setTimeout(() => doHeartbeat(), this.heartbeatIntervalMs)
+        }
+      })
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    if (this.backgroundHeartbeat && window.Worker) {
+      const url = window.URL.createObjectURL(timeoutBlob)
+      worker = new Worker(url)
+      URL.revokeObjectURL(url)
+      worker.onmessage = function () {
+        doHeartbeat()
+      }
+    }
```

Comment:
is the lint exception a false positive?

Reply:
it is. it comes from the fact that it's not aware that you can have one value, do an await (yielding control to a different function), and come back and have that change. It assumes this functions runs all the way through

Reply:
got it - might be helpful to add that as a comment to explain the exception

Reply:
oh i lied. no it's because i have a bool always set to true to use background workers. if we set it to false in the future, it disables this code block here. just easier. ill explain in a comment

Reply:
thanks!